File: algorithms/quicksort.py
import random


def quickSort(nums):
    random.shuffle(nums)
    nums = sortCore(nums, 0, len(nums) - 1)
    return nums


def sortCore(nums, lo, hi):
    if (lo >= hi):
        return
    mi = partition(nums, lo, hi)
    nums = sortCore(nums, lo, mi - 1)
    nums = sortCore(nums, mi + 1, hi)
    return nums


def partition(nums, lo, hi):
    i = lo
    j = hi + 1
    pivot = nums[lo]
    while (True):
        while (pivot > nums[i]):
            i += 1
            if (i == hi):
                break
        while (pivot < nums[j]):
            j -= 1
            if (i >= j):
                break
        nums[i], nums[j] = nums[j], nums[i]
    nums[lo], nums[j] = nums[j], nums[lo]
    return j


def main():
    nums = quickSort([3, 5, 8, 7, 2, 9, 1, 10])
    print(nums)


if __name__ == '__main__':
    main()

File: algorithms/quicksort.cpp
#include<iostream>
#include<vector>

std::vector<int> quickSort(std::vector<int>& nums){
    sortCore(nums, 0, nums.size()-1);
}

void sortCore(std::vector<int> &myNums, int lo, int hi){
    if(lo>hi) return;
    int mi = partition(myNums, lo, mi-1)
    sortCore(myNums, lo, mi-1);
    sortCore(myNums, mi+1, hi);
}

void partition(std::vector<int>& nums, int lo, int hi){
    swap(nums[lo],nums[lo+rand()%(hi-lo+1]));
    int pivot = nums[lo];
    while(lo<hi){
        while(lo<hi){
            if(pivot < nums[hi]) hi--;
            else{
            nums[lo] = nums[hi];
            ++lo;
            break;
            }
        }
        while(lo<hi){
            if(pivot>nums[lo]) lo++;
            else{
                nums[hi] == nums[lo];
                --hi;
                break;
            }
        }
    }
    nums[lo] = pivot;
    return lo;
}